File: flink-libraries/flink-python/src/main/python/org/apache/flink/python/api/flink/functions/ReduceFunction.py
# ###############################################################################
#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
# limitations under the License.
################################################################################
from collections import defaultdict
from flink.functions import Function, RuntimeContext
from flink.connection import Connection, Iterator, Collector


class ReduceFunction(Function.Function):
    def __init__(self):
        super(ReduceFunction, self).__init__()

    def _configure(self, input_file, output_file, port, env, info, subtask_index):
        super(ReduceFunction, self)._configure(input_file, output_file, port, env, info, subtask_index)
        if len(info.key1) == 0:
            self._run = self._run_all_reduce
        else:
            self._run = self._run_grouped_reduce
            self._group_iterator = Iterator.GroupIterator(self._iterator, info.key1)

    def _run(self):
        pass

    def _run_all_reduce(self):
        collector = self._collector
        function = self.reduce
        iterator = self._iterator
        if iterator.has_next():
            base = iterator.next()
            for value in iterator:
                base = function(base, value)
            collector.collect(base)
        collector._close()

    def _run_grouped_reduce(self):
        collector = self._collector
        function = self.reduce
        iterator = self._group_iterator
        iterator._init()
        while iterator.has_group():
            iterator.next_group()
            if iterator.has_next():
                base = iterator.next()
                for value in iterator:
                    base = function(base, value)
            collector.collect(base)
        collector._close()

    def reduce(self, value1, value2):
        pass

    def combine(self, value1, value2):
        return self.reduce(value1, value2)

File: flink-core-api/src/main/java/org/apache/flink/api/common/functions/ReduceFunction.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.flink.api.common.functions;

import org.apache.flink.annotation.Public;

import java.io.Serializable;

/**
 * Base interface for Reduce functions. Reduce functions combine groups of elements to a single
 * value, by taking always two elements and combining them into one. Reduce functions may be used on
 * entire data sets, or on grouped data sets. In the latter case, each group is reduced
 * individually.
 *
 * <p>For a reduce functions that work on an entire group at the same time (such as the
 * MapReduce/Hadoop-style reduce), see {@link GroupReduceFunction}. In the general case,
 * ReduceFunctions are considered faster, because they allow the system to use more efficient
 * execution strategies.
 *
 * <p>The basic syntax for using a grouped ReduceFunction is as follows:
 *
 * <pre>{@code
 * DataSet<X> input = ...;
 *
 * DataSet<X> result = input.groupBy(<key-definition>).reduce(new MyReduceFunction());
 * }</pre>
 *
 * <p>Like all functions, the ReduceFunction needs to be serializable, as defined in {@link
 * java.io.Serializable}.
 *
 * @param <T> Type of the elements that this function processes.
 */
@Public
@FunctionalInterface
public interface ReduceFunction<T> extends Function, Serializable {

    /**
     * The core method of ReduceFunction, combining two values into one value of the same type. The
     * reduce function is consecutively applied to all values of a group until only a single value
     * remains.
     *
     * @param value1 The first value to combine.
     * @param value2 The second value to combine.
     * @return The combined value of both input values.
     * @throws Exception This method may throw exceptions. Throwing an exception will cause the
     *     operation to fail and may trigger recovery.
     */
    T reduce(T value1, T value2) throws Exception;
}
